File: guild_of_heroes/guild_of_heroes_core/src/hero.cpp
#include <stdexcept>
#include "hero.hpp"

std::string
guild_of_heroes::Hero::getName() const {
	return name;
}
int
guild_of_heroes::Hero::getSkill(const std::string skillName) const
{
	auto skillPair = skills.find(skillName);
	if (skillPair == skills.end())
		return 0;
	return skillPair->second;
}
std::map<std::string, int>
guild_of_heroes::Hero::getSkills() const
{
	return skills;
}
int
guild_of_heroes::Hero::getSkillsCount() const {
	return skills.size();
}

int
guild_of_heroes::Hero::getDailyFee() const
{
	return dailyFee;
}

void
guild_of_heroes::Hero::setDailyFee(int fee)
{
	this->dailyFee = fee;
}

void 
guild_of_heroes::Hero::addSkill(const std::string& skillName, int skillValue)
{
	if(skillValue <= 0) 
		throw std::invalid_argument("skillValue must be positive");
	if(skillName.empty())
		throw std::invalid_argument("skillName can not be empty");
	skills[skillName] = skillValue;
}

guild_of_heroes::Hero*
guild_of_heroes::Hero::create(const std::string& name) {
	if (name.empty())
		throw std::invalid_argument("name can not be empty");
	return new Hero(name);
}

guild_of_heroes::Hero::Hero(const std::string& name) {
	this->name = name;
}

File: UnityProject/GuildOfHeroes/Assets/Code/Game/Core/Hero.cs
using GuildOfHeroes.Core.Base;
using System;
using System.Collections.Generic;
using System.Text;

namespace GuildOfHeroes.Core
{
    public class Hero : NameableBase
    {
        public int DailyFee 
        {
            get { return dailyFee; }
            set
            {
                if (value < 0)
                    throw new ArgumentException("Fee can not be negative");
                dailyFee = value;
            }
        }
        public int SkillsCount => skills.Count;
        public IReadOnlyDictionary<string, int> Skills => skills;

        public int GetSkillValue(string skillName)
        {
            if (NotOwnSkill(skillName))
                return 0;
            return skills[skillName];
        }
        public void AddSkill(string skillName, int skillValue)
        {
            if (skillValue <= 0)
                throw new ArgumentException("Skill Value must be positive");
            if (skillName.Length == 0)
                throw new ArgumentException("Skill Name can not be empty");
            if (NotOwnSkill(skillName))
                skills.Add(skillName, skillValue);
            else
                skills[skillName] = skillValue;
        }
        public bool OwnSkill(string skillName)
        {
            return skills.ContainsKey(skillName);
        }
        public bool NotOwnSkill(string skillName)
        {
            return !skills.ContainsKey(skillName);
        }


        public static Hero Create(string name)
        {
            if (name.Length == 0)
                throw new ArgumentException("name can not be empty");
            return new Hero(name);
        }



        private Hero(string name)
            : base(name)
        {
            dailyFee = 0;
            skills = new Dictionary<string, int>();
        }

        private int dailyFee;
        private Dictionary<string, int> skills;
    }
}
